### PR TITLE
Fix issue 1529, Fix validating shader input and output components

### PR DESCRIPTION
**Pull request by ziga-lunarg**
Closes issue #1529

### CODE REVIEW

**Review by ncesario-lunarg — COMMENTED** [review]
I _think_ this looks good, but would be good if a shader guru (@jeremy-lunarg?) could take a peek at the component code.

File: layers/shader_validation.cpp [hunk 1]
```diff
@@ -75,11 +75,41 @@ static bool TypesMatch(SHADER_MODULE_STATE const *a, SHADER_MODULE_STATE const *
         return TypesMatch(a, b, a_insn.word(2), b_type, a_arrayed, b_arrayed, false);
     }
 
+    if (a_insn.opcode() != b_insn.opcode()) {
```
Comment (jeremy-lunarg) [inline layers/shader_validation.cpp:78]: A high level comment for this block would be helpful to future explorers...

**Review by ncesario-lunarg — COMMENTED** [review]
Looks like this is still failing in CI. The shader_module and test changes LGTM, but the `TypesMatch` change is still difficult to follow (for me), and it looks like the failure is happening in `TypesMatch`. Might be worth just refactoring that function rather than making it larger.

File: layers/shader_validation.cpp [hunk 2]
```diff
@@ -2337,35 +2288,55 @@ bool CoreChecks::ValidateInterfaceBetweenStages(SHADER_MODULE_STATE const *produ
...
+        assert(b_at_end || b_component < b_length);
+
         if (b_at_end || ((!a_at_end) && (a_first < b_first))) {
             skip |= LogPerformanceWarning(producer->vk_shader_module(), kVUID_Core_Shader_OutputNotConsumed,
-                                          "%s writes to output location %u.%u which is not consumed by %s", producer_stage->name,
-                                          a_first.first, a_first.second, consumer_stage->name);
-            a_it++;
+                                          "%s writes to output location %" PRIu32 ".%" PRIu32 " which is not consumed by %s",
+                                          producer_stage->name, a_first.first, a_first.second, consumer_stage->name);
+            if (b_first.first > a_first.first || b_at_end || a_component + 1 == a_length) {
```
Comment (ncesario-lunarg) [inline layers/shader_validation.cpp:2313]: Some `()` might help make this a bit clearer.

**Review by ncesario-lunarg — APPROVED** [review]
Nice, thank you for the `TypesMatch` refactor; much easier to read now :)

LGTM once CI passes.

**Pull request merged**